src/tuner/nccl_ofi_regions.c

```diff
@@ -0,0 +1,248 @@
+/*
+ * Copyright (c) 2024 Amazon.com, Inc. or its affiliates. All rights reserved.
+ */
+
+#include "config.h"
+
+#include <math.h>
+
+#include "nccl_ofi_tuner.h"
+
+/* Vector subtraction */
+static inline nccl_ofi_tuner_point_t vsub(nccl_ofi_tuner_point_t a, nccl_ofi_tuner_point_t b)
+{
+	return (nccl_ofi_tuner_point_t){.x = a.x - b.x, .y = a.y - b.y};
+}
+
+/* Vector dot product */
+static inline double vdot(nccl_ofi_tuner_point_t a, nccl_ofi_tuner_point_t b)
+{
+	return a.x * b.x + a.y * b.y;
+}
+
+/* Magnitude of the cross product */
+static inline double vcross(nccl_ofi_tuner_point_t a, nccl_ofi_tuner_point_t b)
+{
+	return a.x * b.y - a.y * b.x;
+}
+
+/* Returns a + s * b */
+static inline nccl_ofi_tuner_point_t vmadd(nccl_ofi_tuner_point_t a, double s, nccl_ofi_tuner_point_t b)
+{
+	nccl_ofi_tuner_point_t c;
+	c.x = a.x + s * b.x;
+	c.y = a.y + s * b.y;
+	return c;
+}
+
+/*
+ * @brief	Check if the x0->x1 segment crosses the y0->y1 segment.
+ *
+ * @param	x0
+ *		point at the beginning of the first segment.
+ * @param	x1
+ *		point at the end of the first segment.
+ * @param	y0
+ *		point at the beginning of the second segment.
+ * @param	y1
+ *		point at the end of the second segment.
+ * @param	eps
+ *		small value used for comparisons.
+ * @param	sect
+ *		If it is a valid pointer, it will be set to the intesection point.
+ *
+ * @return	1 for intersect, -1 for not, 0 if the intersect point
+ *		is too close to y0 or y1.
+ */
+static int intersect(nccl_ofi_tuner_point_t x0,
+		     nccl_ofi_tuner_point_t x1,
+		     nccl_ofi_tuner_point_t y0,
+		     nccl_ofi_tuner_point_t y1,
+		     double eps,
+		     nccl_ofi_tuner_point_t *sect)
+{
+	nccl_ofi_tuner_point_t dx = vsub(x1, x0);
+	nccl_ofi_tuner_point_t dy = vsub(y1, y0);
+	double d = vcross(dy, dx);
+	if (d == 0) {
+		return 0; /* Edges are parallel */
+	}
+
+	double a = (vcross(x0, dx) - vcross(y0, dx)) / d;
+	if (sect) {
+		*sect = vmadd(y0, a, dy);
+	}
+
+	if (a < -eps || a > 1 + eps) {
+		return -1;
+	}
+	if (a < eps || a > 1 - eps) {
+		return 0;
+	}
+
+	a = (vcross(x0, dy) - vcross(y0, dy)) / d;
```

Comment:
If `d` is very small, this could blow up.

Should the above check `d == 0` instead be `abs(d) < eps`?

Reply:
yes, I think you are right